homeassistant/components/google_photos/services.py

```diff
@@ -0,0 +1,104 @@
+"""Google Photos services."""
+
+from __future__ import annotations
+
+import asyncio
+import mimetypes
+from pathlib import Path
+
+import voluptuous as vol
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_FILENAME
+from homeassistant.core import (
+    HomeAssistant,
+    ServiceCall,
+    ServiceResponse,
+    SupportsResponse,
+)
+from homeassistant.exceptions import HomeAssistantError, ServiceValidationError
+from homeassistant.helpers import config_validation as cv
+
+from . import api
+from .const import DOMAIN, UPLOAD_SCOPE
+
+type GooglePhotosConfigEntry = ConfigEntry[api.AsyncConfigEntryAuth]
+
+__all__ = [
+    "DOMAIN",
+]
+
+CONF_CONFIG_ENTRY_ID = "config_entry_id"
+
+UPLOAD_SERVICE = "upload"
+UPLOAD_SERVICE_SCHEMA = vol.Schema(
+    {
+        vol.Required(CONF_CONFIG_ENTRY_ID): cv.string,
+        vol.Required(CONF_FILENAME): vol.All(cv.ensure_list, [cv.string]),
+    }
+)
+
+
+def async_register_services(hass: HomeAssistant) -> None:
+    """Register Google Photos services."""
+
+    async def async_handle_upload(call: ServiceCall) -> ServiceResponse:
+        """Generate content from text and optionally images."""
+        config_entry: GooglePhotosConfigEntry | None = (
+            hass.config_entries.async_get_entry(call.data[CONF_CONFIG_ENTRY_ID])
+        )
+        if not config_entry:
+            raise ServiceValidationError(
+                translation_domain=DOMAIN,
+                translation_key="integration_not_found",
+                translation_placeholders={"target": DOMAIN},
+            )
+        scopes = config_entry.data["token"]["scope"].split(" ")
+        if UPLOAD_SCOPE not in scopes:
+            raise HomeAssistantError(
+                translation_domain=DOMAIN,
+                translation_key="missing_upload_permission",
+                translation_placeholders={"target": DOMAIN},
+            )
+
+        client_api = config_entry.runtime_data
+        upload_tasks = []
+        for filename in call.data[CONF_FILENAME]:
+            if not hass.config.is_allowed_path(filename):
+                raise HomeAssistantError(
+                    translation_domain=DOMAIN,
+                    translation_key="no_access_to_path",
+                    translation_placeholders={"filename": filename},
+                )
+            if not Path(filename).exists():
```

Comment:
I/O inside the event loop. I would move this whole for block inside a single executor job. That way you also don't have to jump between asyncio and executor land if there are multiple files.

Reply:
Fixed.